src/shortwave_radiation.jl

```diff
@@ -0,0 +1,345 @@
+    column.stratz = max(fs0 - column.fsol, 0.0)
+end
+
+"""
+    function cloud!(
+        column::ColumnVariables{NF}, model::PrimitiveEquationModel
+    )
+
+Compute shortwave radiation cloud contibutions for an atmospheric column.
+"""
+function cloud!(
+    column::ColumnVariables{NF}, model::PrimitiveEquationModel
+) where {NF<:AbstractFloat}
+    @unpack rhcl1, rhcl2, rrcl, qcl, pmaxcl, wpcl, gse_s1, gse_s0, clsmax, clsminl =
+        model.parameters
+    @unpack humid, rel_hum, grad_dry_static_energy, precip_convection, precip_large_scale,
+        cloud_top, nlev, fmask = column
+
+    # 1.0 Cloud cover, defined as the sum of:
+    # - a term proportional to the square - root of precip. rate
+    # - a quadratic function of the max. relative humidity
+    #    in tropospheric layers above pbl where q > prog_qcl :
+    #     ( = 0 for rhmax < rhcl1, = 1 for rhmax > rhcl2 )
+    #    Cloud - top level: defined as the highest (i.e. least sigma)
+    #      between the top of convection / condensation and
+    #      the level of maximum relative humidity.
+    if (rel_hum[nlev - 1] > rhcl1)
+        column.cloudc = rel_hum[nlev - 1] - rhcl1
+        column.icltop = nlev - 1
+    else
+        column.cloudc = 0.0
+        column.icltop = nlev + 1
+    end
+
+    for k in 3:(nlev - 2)
+        drh = rel_hum[k] - rhcl1
+        if (drh > column.cloudc) & (humid[k] > qcl)
+            column.cloudc = drh
+            column.icltop = k
+        end
+    end
+
+    pr1 = min(pmaxcl, 86.4 * (precip_convection + precip_large_scale))
+    column.cloudc = min(1.0, wpcl * sqrt(pr1) + min(1.0, column.cloudc * rrcl)^2.0)
+    column.icltop = min(cloud_top, column.icltop)
+
+    # 2.0  Equivalent specific humidity of clouds
+    column.qcloud = humid[nlev - 1]
+
+    # 3.0 Stratiform clouds at the top of pbl
+    clfact = 1.2
+    rgse = 1.0 / (gse_s1 - gse_s0)
+
+    # Stratocumulus clouds over sea
+    fstab = max(0.0, min(1.0, rgse * (grad_dry_static_energy - gse_s0)))
+    column.clstr = fstab * max(clsmax - clfact * column.cloudc, 0.0)
+
+    # Stratocumulus clouds over land
+    clstrl = max(column.clstr, clsminl) * rel_hum[nlev]
+    column.clstr = column.clstr + fmask * (clstrl - column.clstr)
+end
+
+"""
+    function radsw!(
+        column::ColumnVariables{NF}, model::PrimitiveEquationModel
+    )
+
+Compute shortwave radiation fluxes for an atmospheric column.
+"""
+function radsw!(
+    column::ColumnVariables{NF}, model::PrimitiveEquationModel
+) where {NF<:AbstractFloat}
+    @unpack norm_pres,
+    humid, icltop, cloudc, clstr, ozupp, ozone, zenit, stratz, fsol, qcloud, albsfc,
+        nlev = column
+    @unpack σ_levels_full, σ_levels_thick = model.geometry
+
+    @unpack albcl, albcls, abscl1, abscl2, absdry, absaer, abswv1, abswv2, ablwin, ablco2, ablwv1,
+        ablwv2, ablcl2, ablcl1, epslw = model.parameters
+
+    # Locals variables
+    flux::Vector{NF} = fill(NaN, 2)
+    stratc::Vector{NF} = fill(NaN, 2)
```

Comment:
This first creates a `Vector{Float64}` and then converts it to `Vector{NF}` and the previous vector is eventually garbage collected. You can do directly `fill(convert(NF,NaN),2)` instead. However, is it possible to allocate arrays in `column::ColumnVariables`? Or why does a 2-element vector need to be a vector at all?